### PR TITLE
Big refactoring tsc

### PR DESCRIPTION
**Pull request by Arvi3d**

### CODE REVIEW

**Comment by Spoor2709** [comment]
As discussed with Arvid, vsc now finds 1326 issues for elastic search where as the web app and cli find none.

**Comment by Arvi3d** [comment]
> As discussed with Arvid, vsc now finds 1326 issues for elastic search where as the web app and cli find none.

There is an issue in Bundle server. Simon confirmed yesterday. So, this is not related to vscode or tsc

File: src/deepcode/lib/modules/BundlesModule.ts [hunk 1]
```diff
@@ -1,313 +1,90 @@
 import * as vscode from "vscode";
 import * as _ from "lodash";
```
Comment (Spoor2709) [inline src/deepcode/lib/modules/BundlesModule.ts:2]: no longer used

File: src/deepcode/lib/watchers/FilesWatcher.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import * as vscode from 'vscode';
+import { ExtensionInterface } from '../../../interfaces/DeepCodeInterfaces';
+
+import { getGlobPatterns, ISupportedFiles } from '@deepcode/tsc';
+
+export default function createFileWatcher(
+  extension: ExtensionInterface,
```
Comment (Spoor2709) [inline src/deepcode/lib/watchers/FilesWatcher.ts:7]: does extension have to be declared in both functions

File: src/test/suite/extension.test.ts [hunk 3]
```diff
@@ -70,64 +22,60 @@ const preTestConfigureExtension = () => {
...
+  // // set workspace path for tests
+  // testExtension.workspacesPaths = [mockedFolderPath];
+
   return testExtension;
 };
 
-const uri = vscode.Uri.file(
-  path.join(mockedTestFilesDirPath, "../mocked_data/sample_repository", "main.js"),
-);
+const uri = vscode.Uri.file(nodePath.join(mockedTestFilesDirPath, '../mocked_data/sample_repository', 'main.js'));
```
Comment (Spoor2709) [inline src/test/suite/extension.test.ts:32]: not used

File: src/test/suite/extension.test.ts [hunk 4]
```diff
@@ -70,64 +22,60 @@ const preTestConfigureExtension = () => {
...
 
 const testIgnoreComment = '  // deepcode ignore UseStrictEquality: <please specify a reason of ignoring this>\n';
```
Comment (Spoor2709) [inline src/test/suite/extension.test.ts:34]: not used

File: src/test/suite/extension.test.ts [hunk 5]
```diff
@@ -70,64 +22,60 @@ const preTestConfigureExtension = () => {
...
-    const editor = await vscode.window.showTextDocument(document, 1, false);
-    return editor.edit(textEditor => {
-      textEditor.insert(new vscode.Position(18, 0), testIgnoreComment);
-    }).then(inserted => {
-      assert.equal(`${document.lineAt(18).text}\n`, testIgnoreComment);
-      // TODO: find a way to undo this change
-      // TODO: check actual analysis results with ignored line
-    });
-  });
+  // This is useless test
```
[the remark this answers is not in the document]
Comment (Spoor2709) [reply src/test/suite/extension.test.ts:48]: Maybe add appropriate tests?

**Review by Spoor2709 — COMMENTED** [review]
Looks really good, a very stable version :) Tested on windows, mac and linux. Also It looks like src/deepcode/utils/throttle.ts is not longer used

File: src/deepcode/lib/modules/LoginModule.ts [hunk 6]
```diff
@@ -23,31 +26,38 @@ abstract class LoginModule extends ReportModule implements DeepCode.LoginModuleI
     try {
       const checkCurrentToken = await this.checkSession();
       if (checkCurrentToken) return;
-      const result = await http.login(this.baseURL, this.source);
-      const { sessionToken, loginURL } = result;
-      if (!sessionToken || !loginURL) {
+      const response = await startSession({ baseURL: this.baseURL, source: this.source });
```
Comment (shinin-light) [inline src/deepcode/lib/modules/LoginModule.ts:29]: This new flow breaks a fundamental user interaction:
1. The user clicks on the login button
2. The browser opens for the login (the token is not recorded yet)
3. The user takes more than 20 seconds to onboard and login (the waitLoginConfirmation exits with no success)
4. The user is told on the browser to close the tab and go back to VSC
5. The user still sees the login button

-> by clicking on it, in the previous flow would just check the token and move to the accept-upload step. Now however the flow goes back to point 2, creating a new token and everything else. The previously created token is lost.

**Review by shinin-light — CHANGES_REQUESTED** [review]
Small login and disposables issues.

File: src/test/suite/extension.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/test/suite/extension.test.ts:48]: yes, but this is going to be a separate task.

File: src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts [hunk 7]
```diff
@@ -1,111 +1,94 @@
...
-    this.createReviewResults();
+    this.extension.processEvent(TELEMETRY_EVENTS.ignoreSuggestion, options);
   }
 
   private createIssueDiagnosticInfo({
-    issue,
     issuePositions,
-    suggestions,
-    fileUri
+    suggestion,
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts:66]: I did not catch here.

File: src/deepcode/lib/modules/BaseDeepCodeModule.ts [hunk 8]
```diff
@@ -193,17 +185,19 @@ export default abstract class BaseDeepCodeModule implements DeepCode.BaseDeepCod
...
+          ),
         )
-      ).then(
-        () => {},
-        (error) => self.processError(error, {
-          message: errorsLogs.loadingBadge,
-        })
-      );
+        .then(
+          () => {},
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/lib/modules/BaseDeepCodeModule.ts:195]: i'd change this with "catch", but i did not touch this function. it was there already like this

File: src/deepcode/DeepCodeExtension.ts [hunk 9]
```diff
@@ -14,25 +17,15 @@ import {
...
-  ): Promise<any> {
+} from './constants/commands';
+import { openDeepcodeSettingsCommand, createDCIgnoreCommand } from './utils/vscodeCommandsUtils';
+import { errorsLogs } from './messages/errorsServerLogMessages';
+import { DEEPCODE_VIEW_SUPPORT, DEEPCODE_VIEW_ANALYSIS, DEEPCODE_ANALYSIS_STATUS } from './constants/views';
+import { SupportProvider } from './view/SupportProvider';
+import { IssueProvider } from './view/IssueProvider';
+
+class DeepCodeExtension extends DeepCodeLib implements ExtensionInterface {
+  private async executeCommand(name: string, fn: (...args: any[]) => Promise<any>, ...args: any[]): Promise<any> {
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/DeepCodeExtension.ts:28]: yep, i did not touch it

File: src/deepcode/DeepCodeExtension.ts [hunk 10]
```diff
@@ -131,54 +106,60 @@ class DeepCodeExtension extends DeepCodeLib implements DeepCode.ExtensionInterfa
...
-      new SupportProvider(this)
-    );
+    vscode.window.registerTreeDataProvider(DEEPCODE_VIEW_SUPPORT, new SupportProvider(this));
+
+    vscode.window.registerTreeDataProvider(DEEPCODE_VIEW_ANALYSIS, new IssueProvider(this));
 
-    vscode.window.registerTreeDataProvider(
-      DEEPCODE_VIEW_ANALYSIS, 
-      new IssueProvider(this)
+    vscode.workspace.onDidChangeWorkspaceFolders(() => this.startExtension());
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/DeepCodeExtension.ts:128]: nope, it's not required here

File: src/deepcode/lib/modules/LoginModule.ts [hunk 11]
```diff
@@ -23,31 +26,38 @@ abstract class LoginModule extends ReportModule implements DeepCode.LoginModuleI
...
   }
 
-  async checkSession(): Promise<boolean> {
+  async checkSession(token = ''): Promise<boolean> {
     let validSession = false;
-    if (this.token) {
+    if (token || this.token) {
       try {
-        validSession = !!(await http.checkSession(this.baseURL, this.token));
+        validSession = !!(await checkSession({
```
Comment (shinin-light) [inline src/deepcode/lib/modules/LoginModule.ts:54]: This also doesn't work anymore with the current tsc response (it returns true even for 304 errors). It should be something like:

`const session = await checkSession({
          baseURL: this.baseURL,
          sessionToken: token || this.token,
        });
        if (session.type === 'error') {
          validSession = false;
        } else {
          validSession = session.value;
        }`

File: src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts [hunk 12]
```diff
@@ -161,40 +132,34 @@ class DeepCodeAnalyzer implements DeepCode.AnalyzerInterface {
...
     }
+
     this.setIssuesMarkersDecoration();
   }
 
-  public async configureIssuesDisplayBySeverity(
-    severity: number,
-    hide: boolean
-  ): Promise<void> {
+  public async configureIssuesDisplayBySeverity(severity: number, hide: boolean): Promise<void> {
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts]: done

File: src/deepcode/DeepCodeExtension.ts [hunk 13]
```diff
@@ -1,8 +1,11 @@
-import * as vscode from "vscode";
-import * as open from "open";
+import * as vscode from 'vscode';
+import open from 'open';
+import { emitter } from '@deepcode/tsc';
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/DeepCodeExtension.ts]: done

File: src/deepcode/lib/deepCodeProviders/hoverProvider/DeepCodeHoverProvider.ts [hunk 14]
```diff
@@ -15,7 +15,7 @@ export class DisposableHoverProvider implements vscode.Disposable {
     this.hoverProvider = vscode.languages.registerHoverProvider(
       { scheme: "file", language: "*" },
       {
-        provideHover(document, position, token) {
+        provideHover(document, position, _token) {
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/lib/deepCodeProviders/hoverProvider/DeepCodeHoverProvider.ts]: done

File: src/deepcode/lib/modules/BundlesModule.ts (hunk 1 above)
Comment (Arvi3d) [reply src/deepcode/lib/modules/BundlesModule.ts]: done

File: src/deepcode/lib/watchers/FilesWatcher.ts (hunk 2 above)
Comment (Arvi3d) [reply src/deepcode/lib/watchers/FilesWatcher.ts:7]: done

File: src/test/suite/extension.test.ts (hunk 4 above)
Comment (Arvi3d) [reply src/test/suite/extension.test.ts]: done

File: src/test/suite/extension.test.ts (hunk 3 above)
Comment (Arvi3d) [reply src/test/suite/extension.test.ts]: done

File: src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts [hunk 15]
```diff
@@ -1,111 +1,94 @@
...
-    this.analysisResultsCollection = {};
-    this.ignoreActionsProvider = new DisposableCodeActionsProvider(
-      this.deepcodeReview,
-      {
-        findSuggestionId: this.findSuggestionId.bind(this),
-        trackIgnoreSuggestion: this.trackIgnoreSuggestion.bind(this)
-      }
-    );
-    this.issueHoverProvider = new DisposableHoverProvider(this.deepcodeReview);
+    new DisposableCodeActionsProvider(this.deepcodeReview, {
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts]: reverted as it used to be

File: src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts [hunk 16]
```diff
@@ -1,111 +1,94 @@
...
+      findSuggestion: this.findSuggestion.bind(this),
+      trackIgnoreSuggestion: this.trackIgnoreSuggestion.bind(this),
+    });
+    new DisposableHoverProvider(this.deepcodeReview);
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/lib/analyzer/DeepCodeAnalyzer.ts]: done

File: src/deepcode/DeepCodeExtension.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Arvi3d) [reply src/deepcode/DeepCodeExtension.ts]: refactored for consistency

File: src/deepcode/lib/modules/LoginModule.ts (hunk 11 above)
Comment (Arvi3d) [reply src/deepcode/lib/modules/LoginModule.ts]: done

File: src/deepcode/lib/modules/LoginModule.ts (hunk 6 above)
Comment (Arvi3d) [reply src/deepcode/lib/modules/LoginModule.ts:29]: Yes, this is possible, but still i don't see here a better way to handle login process.
Previously, we were saving a token to the extension, that is not confirmed. This caused many problems.
I believe, we have to save only confirmed tokens. It also simplifies following logics, when we expect token to be confirmed when we use it, with just handling an error if it's not.

Comment (Arvi3d) [reply src/deepcode/lib/modules/LoginModule.ts:44]: done

**Review by Spoor2709 — APPROVED** [review]

File: src/deepcode/lib/modules/LoginModule.ts [hunk 17]
```diff
@@ -23,31 +27,59 @@ abstract class LoginModule extends ReportModule implements DeepCode.LoginModuleI
...
+            await this.setToken(this.pendingToken);
+            return;
+          }
+        } finally {
+          this.pendingToken = '';
+        }
+      }
+
+      const response = await startSession({ baseURL: this.baseURL, source: this.source });
+      if (response.type == 'error') {
```
Comment (shinin-light) [inline src/deepcode/lib/modules/LoginModule.ts:45]: Valid, but please use === instead of ==

**Review by shinin-light — APPROVED** [review]

File: package.json [hunk 18]
```diff
@@ -220,38 +221,37 @@
...
     "eslint-plugin-prettier": "^3.1.4",
-    "eslint-plugin-promise": "^4.2.1",
-    "eslint-plugin-standard": "^4.0.1",
+    "glob": "^7.1.6",
+    "mocha": "^8.1.3",
+    "prettier": "^2.1.1",
+    "typescript": "^4.0.2",
     "vscode-test": "^1.4.0",
-    "nock": "^12.0.3"
+    "yalc": "^1.0.0-pre.44"
```
[the remark this answers is not in the document]
Comment (Arvi3d) [reply package.json:250]: it's in devDependencies and is needed in the future

File: src/deepcode/lib/modules/LoginModule.ts (hunk 17 above)
Comment (Arvi3d) [reply src/deepcode/lib/modules/LoginModule.ts]: done

**Pull request merged**